java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.expression.Expression
+
+internal typealias Path = List<AccessPath.Selector.Field>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+/** A visitor that accumulates field [Paths] from a Paxel [Expression]. */
+class ExpressionPathAccumulator : Expression.Visitor<Paths> {
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>): Paths {
+        return expr.left.accept(this) + expr.right.accept(this)
+    }
+
+    override fun <T> visit(expr: Expression.FieldExpression<T>): Paths {
+        if (expr.qualifier == null) {
+            return listOf(listOf(AccessPath.Selector.Field(expr.field)))
+        }
+        return listOf(
+            expr.qualifier!!.accept(this).first() + listOf(AccessPath.Selector.Field(expr.field))
```

Comment:
Don't you need to create a cross product between qualiifer fields and the current field?  e.g., if visiting the qualifier returns `[a, b]` and the field is `f`, shouldn't the result be `[a.f, b.f]`?

Reply:
This is an interesting concern. I'm not sure if we need to implement a cross, but maybe I'm missing something. From my understanding, there are three cases that need to be handled -- this is reflected in the `when` statement in the current (2020-09-23) implementation.